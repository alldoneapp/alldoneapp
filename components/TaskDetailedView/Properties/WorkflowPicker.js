import React, { useRef, useState } from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { colors } from '../../styles/global'
import Popover from 'react-tiny-popover'
import StatusPicker from './StatusPicker'
import { hideFloatPopup, showFloatPopup } from '../../../redux/actions'
import TasksHelper, { DONE_STEP, OPEN_STEP } from '../../TaskListView/Utils/TasksHelper'
import Button from '../../UIControls/Button'
import Icon from '../../Icon'
import { chronoEntriesOrder } from '../../../utils/HelperFunctions'
import { useDispatch, useSelector } from 'react-redux'
import { WORKSTREAM_ID_PREFIX } from '../../Workstreams/WorkstreamHelper'
import { translate } from '../../../i18n/TranslationService'

const WorkflowPicker = ({ projectId, task, disabled }) => {
    const [visiblePopover, setVisiblePopover] = useState(false)
    const assignee = useSelector(state => state.assignee)
    const loggedUser = useSelector(state => state.loggedUser)
    const smallScreen = useSelector(state => state.smallScreen)
    const dispatch = useDispatch()
    const ownerIsWorkstream = task?.userId?.startsWith(WORKSTREAM_ID_PREFIX)
    const btnRef = useRef()

    const getStepData = workflow => {
        const workflowEntries = workflow ? Object.entries(workflow).sort(chronoEntriesOrder) : []
        let stepNumber = OPEN_STEP

        if (task.done) {
            stepNumber = DONE_STEP
        } else {
            for (let i = 0; i < workflowEntries.length; ++i) {
                if (workflowEntries[i][0] === task.stepHistory[task.stepHistory.length - 1]) {
                    stepNumber = i
                    break
                }
            }
        }

        switch (stepNumber) {
            case OPEN_STEP:
                return { number: OPEN_STEP, name: translate('Open') }
            case DONE_STEP:
                return { number: DONE_STEP, name: translate('Done') }
            default:
                return {
                    number: stepNumber + 1,
                    name: workflowEntries[stepNumber][1].description,
                }
        }
    }

    const getWorkflow = () => {
        let taskOwner = ownerIsWorkstream ? loggedUser : assignee || TasksHelper.getTaskOwner(task.userId, projectId)

        if (!taskOwner.workflow) {
            taskOwner.workflow = {}
        }

        return taskOwner.workflow[projectId] ? taskOwner.workflow[projectId] : {}
    }

    const showPopover = () => {
        setVisiblePopover(true)
        dispatch(showFloatPopup())
        btnRef?.current?.blur()
    }

    const hidePopover = () => {
        setVisiblePopover(false)
        dispatch(hideFloatPopup())
    }

    const workflow = getWorkflow()
    const stepData = getStepData(workflow)

    return (
        <Popover
            content={<StatusPicker workflow={workflow} projectId={projectId} task={task} hidePopover={hidePopover} />}
            onClickOutside={hidePopover}
            isOpen={visiblePopover}
            position={['bottom', 'left', 'right', 'top']}
            padding={4}
            align={'end'}
            contentLocation={smallScreen ? null : undefined}
        >
            <Button
                ref={btnRef}
                icon={
                    stepData.number === DONE_STEP ? (
                        <Icon name={'square-checked-gray'} size={20} color={colors.Text03} />
                    ) : (
                        <View style={localStyles.stepMarker}>
                            <Text style={localStyles.stepNumber}>
                                {stepData.number === OPEN_STEP ? '' : stepData.number}
                            </Text>
                        </View>
                    )
                }
                title={stepData.name}
                type={'ghost'}
                onPress={showPopover}
                disabled={disabled}
            />
        </Popover>
    )
}
export default WorkflowPicker

const localStyles = StyleSheet.create({
    stepMarker: {
        borderColor: colors.Text03,
        borderRadius: 4,
        borderWidth: 2,
        alignItems: 'center',
        justifyContent: 'center',
        height: 20,
        width: 20,
    },
    stepNumber: {
        color: colors.Text03,
        fontFamily: 'Roboto-Regular',
        fontSize: 10,
    },
})
