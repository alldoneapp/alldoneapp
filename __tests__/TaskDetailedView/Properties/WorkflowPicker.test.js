/**
 * @jest-environment jsdom
 */

import React from 'react'
import WorkflowPicker from '../../../components/TaskDetailedView/Properties/WorkflowPicker'
import store from '../../../redux/store'
import { storeCurrentUser } from '../../../redux/actions'

import renderer from 'react-test-renderer'

jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn().mockImplementation(fnc => {
        return fnc({
            currentUser: {},
            assignee: { workflow: {} }
        })
    }),
    useDispatch: jest.fn(),
    useStore: jest.fn().mockImplementation(() => {
        return {
            getState: () => {
                return { selectedProjectIndex: 0, projectsUsers: [[]] }
            }
        }
    })
}));

describe('WorkflowPicker component', () => {
    describe('WorkflowPicker snapshot test', () => {
        it('should render correctly', () => {
            store.dispatch(storeCurrentUser({workflow:[]}))
            const tree = renderer.create(<WorkflowPicker task={{ id: '0', done: false, inReview: false, toReview: 0 }}/>).toJSON()
            expect(tree).toMatchSnapshot()
        })
    })
})
