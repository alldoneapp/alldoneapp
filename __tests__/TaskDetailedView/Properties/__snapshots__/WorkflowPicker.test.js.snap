// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowPicker component WorkflowPicker snapshot test should render correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "flex-start",
      "backgroundColor": "transparent",
      "borderColor": "#C6CDD2",
      "borderRadius": 4,
      "borderStyle": "solid",
      "borderWidth": 1,
      "flexDirection": "row",
      "flexWrap": "nowrap",
      "height": 40,
      "justifyContent": "center",
      "maxHeight": 40,
      "minHeight": 40,
      "opacity": 1,
      "paddingLeft": 12,
      "paddingRight": 16,
      "paddingVertical": 8,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#8A94A6",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#8A94A6",
          "fontFamily": "Roboto-Regular",
          "fontSize": 10,
        }
      }
    >
      
    </Text>
  </View>
  <Text
    numberOfLines={1}
    style={
      Array [
        Array [
          Object {
            "alignSelf": "center",
            "color": "#FFFFFF",
            "flexWrap": "nowrap",
            "fontFamily": "Roboto-Medium",
            "fontSize": 14,
            "letterSpacing": 0.8,
            "lineHeight": 16,
            "margin": 0,
            "paddingHorizontal": 0,
            "paddingVertical": 0,
          },
          Object {
            "color": "#8A94A6",
          },
          Object {},
        ],
        Object {
          "marginLeft": 12,
        },
      ]
    }
  >
    Open
  </Text>
</View>
`;
